### DISCUSSION TITLE
Running Argo workflows from a remote machine on Nebari

### DISCUSSION THREAD

**Discussion by rsignell** [nebari-dev/nebari#2183] [Q&A]
I tried out using Argo from the Argo UI, and it worked!  Awesome!

Now I'm wondering whether it is possible to launch an Argo workflow on the Nebari kube cluster from a remote machine (e.g. my laptop) via CLI commands and still be able to use existing conda environments via the [Nebari Workflow Controller](https://www.nebari.dev/docs/how-tos/using-argo#access-your-nebari-environments-and-file-system-while-on-an-argo-pod-beta)?

Aside:  Although I couldn't figure out how to run Argo workflows on the Nebari k8s from my laptop, I did manage to launch jobs there using [SkyPilot](https://github.com/OpenScienceComputing/opendrift-sky/blob/main/README.md).  I couldn't take advantage of any of the Nebari goodies like conda-store environments of course.

**Comment by dharhas** [comment]
I'm pretty sure I have done this a while back. If you setup credentials correctly there shouldn't be anything stopping you. I'll try and look up my old notes.

The conda-store env availability is of course the biggest usability issue. There is a way to make the conda env available via argo workflow controller (https://www.nebari.dev/docs/how-tos/setup-argo/) but it is not well documented on how to use it right now. @kcpevey was doing some experiments with Hera and was struggling with activating environments etc.

**Comment by kcpevey** [comment]
The outcome of my struggles with argo/hera is summarized [here](https://www.nebari.dev/docs/tutorials/argo-workflows-walkthrough/#argo-assistant-code) in the docs. I ended up creating some helper functions (included) which help you construct a single line command to execute a python script in a conda environment via the argo scheduler.

**Reply by rsignell** [reply]
@kcpevey, just to make sure I understand,  with your helper script I should be able to run:
``` python
path = '/path/to/pyfile.py'
nebari_conda_env = 'analyst-workflow-env'
submit_argo_script(path, nebari_conda_env)
```
**without being logged into Nebari** (e.g. running Python on my local laptop), right?

If so, do I also need to include this in the helper script:
``` python
from hera.shared import global_config
import os

global_config_host = f"https://{os.environ['ARGO_SERVER'].rsplit(':')[0]}{os.environ['ARGO_BASE_HREF']}/"
global_config.host = global_config_host
global_config.token = os.environ['ARGO_TOKEN']
```
If so, where do I get those `ARGO_TOKEN` , `ARGO_BASE_HREF` and `ARGO_SERVER` values from that will work on my local machine?

**Reply by maxrjones** [reply]
Hey @rsignell, did you find a convenient way to get the `ARGO_TOKEN`? For submitting argo workflows via the cli from JupyterLab, I have been copying the token to clipboard from `https://<your-nebari-url>/argo/userinfo`. But I can't paste the entire section since the slash in `ARGO_BASE_HREF` needs to be in a different spot than in the copied version (xref https://github.com/nebari-dev/nebari-docker-images/issues/83#issuecomment-1658736014). This method is inconvenient to do each time the token expires, so I'm hoping there's a better way.

**Reply by rsignell** [reply]
@maxrjones no I never even figured out how to run an Argo workflow without logging into Nebari, which kind of defeats the purpose. 🥲

**Reply by dharhas** [reply]
@maxrjones the only part that changes is the token, so why not copy that instead of the entire section? The rest of it can stay static in a .env file or similar if using something like dotenv.

@rsignell you have to log into Nebari now and then to get a token otherwise anyone on the internet would be able to run things. Now the question becomes what the token expiry length is set to.

**Reply by maxrjones** [reply]
> @maxrjones the only part that changes is the token, so why not copy that instead of the entire section? The rest of it can stay static in a .env file or similar if using something like dotenv.

When I go to `https://<your-nebari-url>/argo/userinfo`, that line appears as `export ARGO_TOKEN=''`, so I don't know how to access the token without selecting copy to clipboard (which copies the whole block). I would love to know if I'm missing something obvious about how to copy just that line.

> @maxrjones no I never even figured out how to run an Argo workflow without logging into Nebari, which kind of defeats the purpose. 🥲

Ah, sorry to hear that. We're trying to use it to run many models at once, so I'm hoping it still works well for parallelizing runs even though they need to be submitting via the JupyterLab interface.

**Reply by kcpevey** [reply]
@maxrjones another option is to go through the terminal on jupyterlab and pull it from the environment variables there `echo $ARGO_TOKEN`. That would also be available via notebooks in the jupyterLab interface (e.g. `os.environ['ARGO_TOKEN']`).

**Reply by dharhas** [reply]
The other option is whether Argo settings can be modified to tweak that clipboard item. I have no idea if that is possible.

**Reply by maxrjones** [reply]
> @maxrjones another option is to go through the terminal on jupyterlab and pull it from the environment variables there `echo $ARGO_TOKEN`. That would also be available via notebooks in the jupyterLab interface (e.g. `os.environ['ARGO_TOKEN']`).

Hmm, interesting. While `echo $ARGO_TOKEN` returns a token, when I run `argo list` via the JupyterLab without copying the token from the Argo dashboard, I get `rpc error: code = Unauthenticated desc = failed to validate claims: go-jose/go-jose/jwt: validation failed, token is expired (exp)`. Should I open an issue about this if it isn't expected behavior?

**Reply by kcpevey** [reply]
It sounds like the argo cli isn't using that ARGO_TOKEN. Have you tried to feed the token directly to the argo cli to authenticate? I've only used the cli a few times a while back so I don't know much about it.

**Reply by Adam-D-Lewis** [reply]
@maxrjones I looked into this a bit and found that The ARGO_TOKEN set by default works if you modify it to add `Bearer ` (Bearer then a space, also it's case-sensitive) before what it is currently set to.  That is a bug.

**Reply by Adam-D-Lewis** [reply]
Looking into it a bit more, it's actually that way b/c hera (python client for argo workflows) expects the token without "Bearer " and it was designed for that use case.  I'll propose that we set a different env var `HERA_TOKEN` for use with Hera in an issue.

**Reply by Adam-D-Lewis** [reply]
Opened an issue: https://github.com/nebari-dev/nebari/issues/2437

**Reply by maxrjones** [reply]
Thank you @Adam-D-Lewis for figuring that out! Updating `ARGO_TOKEN` to include `Bearer ` would make using argo via Nebari so much more convenient for us, so very excited about https://github.com/nebari-dev/nebari/pull/2438 🎉
